File: src/main/java/me/aov/sellgui/listeners/InventoryListeners.java
package me.aov.sellgui.listeners;

import me.aov.sellgui.SellGUI;
import me.aov.sellgui.SellGUIMain;
import me.aov.sellgui.commands.SellCommand;
import me.aov.sellgui.handlers.SoundHandler;
import org.bukkit.*;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.inventory.*;
import org.bukkit.persistence.PersistentDataType;

import java.util.HashMap;

public class InventoryListeners implements Listener {
    private final SellGUIMain main;

    public InventoryListeners(SellGUIMain main) {
        this.main = main;
    }

    @EventHandler
    public void onInventoryClose(InventoryCloseEvent event) {
        if (!(event.getPlayer() instanceof Player)) return;

        Player player = (Player) event.getPlayer();
        SellGUI sellGUI = SellCommand.getSellGUI(player);

        if (sellGUI != null && event.getInventory().equals(sellGUI.getMenu())) {

            dropItems(event.getInventory(), player);

            sellGUI.cleanup();
            SellCommand.getSellGUIs().remove(sellGUI);
        }
    }

    @EventHandler
    public void onInventoryClick(InventoryClickEvent e) {
        if (!(e.getWhoClicked() instanceof Player)) return;

        Player player = (Player) e.getWhoClicked();
        Inventory clickedInventory = e.getClickedInventory();
        ItemStack currentItem = e.getCurrentItem();

        SellGUI sellGUI = SellCommand.getSellGUI(player);
        if (sellGUI == null || !e.getView().getTopInventory().equals(sellGUI.getMenu())) {
            return;
        }

        NamespacedKey guiKey = new NamespacedKey(main, "sellgui");
        NamespacedKey actionKey = new NamespacedKey(main, "guiAction");

        if (currentItem != null && currentItem.hasItemMeta() &&
                currentItem.getItemMeta().getPersistentDataContainer().has(guiKey, PersistentDataType.BYTE)) {

            e.setCancelled(true);

            String action = currentItem.getItemMeta().getPersistentDataContainer().get(actionKey, PersistentDataType.STRING);
            if (action == null) return;

            switch (action) {
                case "confirm":
                    if (sellGUI.hasUnevaluatedItems()) {
                        player.closeInventory();
                        String message = main.getMessagesConfig().getString("messages.evaluation-required", "&cYou have items that need to be evaluated before selling.");
                        player.sendMessage(ChatColor.translateAlternateColorCodes('&', message));
                        return;
                    }
                    SoundHandler.playConfigSound(player, "sounds.ui.confirm");
                    sellGUI.sellItems(sellGUI.getMenu());
                    break;
                case "sell":
                    if (sellGUI.getTotal(sellGUI.getMenu()) <= 0) {

                        String soundName = main.getSoundsConfig().getString("legacy.no-items-sound", "BLOCK_NOTE_BLOCK_BASS");
                        float volume = (float) main.getSoundsConfig().getDouble("legacy.no-items-volume", 1.0);
                        float pitch = (float) main.getSoundsConfig().getDouble("legacy.no-items-pitch", 1.0);

                        try {
                            Sound sound = Sound.valueOf(soundName);
                            player.playSound(player.getLocation(), sound, volume, pitch);
                        } catch (IllegalArgumentException ex) {
                            main.getLogger().warning("Invalid sound name: " + soundName);
                        }

                        player.closeInventory();
                        player.sendTitle(
                                main.getMessagesConfig().getString("titles.no-items-title", "&cNo items to sell!"),
                                main.getMessagesConfig().getString("titles.no-items-subtitle", ""),
                                10, 70, 20
                        );
                        return;
                    }
                    SoundHandler.playUIClick(player);
                    sellGUI.makeConfirmItem();
                    sellGUI.setConfirmItem();
                    break;
            }
        }

        else if (currentItem != null && isCustomMenuItem(currentItem)) {
            e.setCancelled(true);
            handleCustomMenuItemClick(player, currentItem);
        }

        else if (clickedInventory != null && clickedInventory.equals(sellGUI.getMenu())) {

            if (isGUIControlItem(currentItem, sellGUI)) {
                e.setCancelled(true);
                return;
            }

            if (e.isShiftClick() && wouldGoToReservedSlot(e.getSlot(), sellGUI)) {
                e.setCancelled(true);
                return;
            }

            if (e.getCursor() != null && e.getCursor().getType() != Material.AIR &&
                    (e.getCurrentItem() == null || e.getCurrentItem().getType() == Material.AIR)) {
                SoundHandler.playConfigSound(player, "sounds.items.place");
            }

            if (!e.isCancelled()) {
                SoundHandler.playUIClick(player);
            }

            Bukkit.getScheduler().runTaskLater(main, () -> {
                if (sellGUI.getMenu() != null && player.isOnline() && SellCommand.getSellGUI(player) != null) {
                    sellGUI.updateButtonState();
                }
            }, 1L);
        }

        else if (clickedInventory != null && clickedInventory.equals(player.getInventory())) {

            if (e.isShiftClick() && currentItem != null && currentItem.getType() != Material.AIR) {
                SoundHandler.playConfigSound(player, "sounds.items.place");
            }

            Bukkit.getScheduler().runTaskLater(main, () -> {
                if (sellGUI.getMenu() != null && player.isOnline() && SellCommand.getSellGUI(player) != null) {
                    sellGUI.updateButtonState();
                }
            }, 1L);
        }
    }

    private boolean isGUIControlItem(ItemStack item, SellGUI sellGUI) {
        if (item == null) return false;

        NamespacedKey guiKey = new NamespacedKey(main, "sellgui");
        return item.hasItemMeta() &&
                item.getItemMeta().getPersistentDataContainer().has(guiKey, PersistentDataType.BYTE);
    }

    private boolean wouldGoToReservedSlot(int slot, SellGUI sellGUI) {

        if (main.getConfigManager() != null && main.getConfigManager().getGUIConfig() != null) {
            var guiConfig = main.getConfigManager().getGUIConfig();
            var fillerSlots = guiConfig.getIntegerList("sell_gui.positions.filler_slots");
            int sellButtonSlot = guiConfig.getInt("sell_gui.positions.sell-button", 49);
            int confirmButtonSlot = guiConfig.getInt("sell_gui.positions.confirm_button", 53);

            return fillerSlots.contains(slot) || slot == sellButtonSlot || slot == confirmButtonSlot;
        }

        return slot < 9 || slot > 44 || slot % 9 == 0 || slot % 9 == 8;
    }

    private void handleCustomMenuItemClick(Player player, ItemStack item) {
        NamespacedKey key = new NamespacedKey(main, "custom-menu-item");
        String commands = item.getItemMeta().getPersistentDataContainer().get(key, PersistentDataType.STRING);

        if (commands != null && !commands.isEmpty()) {
            String[] commandArray = commands.split(";");
            for (String command : commandArray) {
                if (!command.trim().isEmpty()) {

                    String finalCommand = command.trim().replace("%player%", player.getName());
                    Bukkit.dispatchCommand(Bukkit.getConsoleSender(), finalCommand);
                }
            }
        }
    }

    private void dropItems(Inventory inventory, Player player) {
        for (ItemStack itemStack : inventory.getContents()) {
            if (itemStack == null || itemStack.getType() == Material.AIR) continue;

            NamespacedKey guiKey = new NamespacedKey(main, "sellgui");
            if (itemStack.hasItemMeta() &&
                    itemStack.getItemMeta().getPersistentDataContainer().has(guiKey, PersistentDataType.BYTE)) {
                continue;
            }

            if (isCustomMenuItem(itemStack)) {
                continue;
            }

            if (main.getConfig().getBoolean("drop-items-on-close", false)) {
                player.getWorld().dropItem(player.getLocation(), itemStack);
            } else {
                HashMap<Integer, ItemStack> notAdded = player.getInventory().addItem(itemStack);
                if (!notAdded.isEmpty()) {
                    for (ItemStack leftover : notAdded.values()) {
                        player.getWorld().dropItem(player.getLocation(), leftover);
                    }
                }
            }
        }
    }

    private boolean isCustomMenuItem(ItemStack item) {
        if (item == null || !item.hasItemMeta()) return false;
        return item.getItemMeta().getPersistentDataContainer().has(
                new NamespacedKey(main, "custom-menu-item"),
                PersistentDataType.STRING
        );
    }

    public static boolean sellGUIItem(ItemStack item, Player player) {
        if (item == null || !item.hasItemMeta()) return false;

        SellGUIMain instance = SellGUIMain.getInstance();
        if (instance == null) return false;

        NamespacedKey guiKey = new NamespacedKey(instance, "sellgui");
        return item.getItemMeta().getPersistentDataContainer().has(guiKey, PersistentDataType.BYTE);
    }
}

File: src/main/kotlin/me/aov/sellgui/listeners/InventoryListeners.kt
package me.aov.sellgui.listeners

import me.aov.sellgui.SellGUI
import me.aov.sellgui.SellGUIMain
import me.aov.sellgui.commands.SellCommand
import me.aov.sellgui.handlers.SoundHandler
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryAction
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.inventory.InventoryDragEvent
import org.bukkit.event.player.PlayerDropItemEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType

class InventoryListeners(private val main: SellGUIMain) : Listener {
    @EventHandler
    fun onInventoryClose(event: InventoryCloseEvent) {
        val player = event.player as? Player ?: return
        val sellGUI = SellCommand.getSellGUI(player) ?: return
        if (event.inventory != sellGUI.getMenu()) return
        dropItems(event.inventory, player)
        sellGUI.cleanup()
        SellCommand.getSellGUIs().remove(sellGUI)
    }

    @EventHandler
    fun onInventoryDrag(event: InventoryDragEvent) {
        val player = event.whoClicked as? Player ?: return
        val sellGUI = SellCommand.getSellGUI(player) ?: return
        if (event.view.topInventory != sellGUI.getMenu()) return
        for (slot in event.rawSlots) {
            if (slot < event.view.topInventory.size) {
                val item = event.view.topInventory.getItem(slot)
                if (isGUIControlItem(item) || isCustomMenuItem(item)) {
                    event.isCancelled = true
                    return
                }
            }
        }
        scheduleMenuRefresh(player, sellGUI)
    }

    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        val player = event.whoClicked as? Player ?: return
        val currentItem = event.currentItem
        val sellGUI = SellCommand.getSellGUI(player) ?: return
        if (event.view.topInventory != sellGUI.getMenu()) return
        if (event.action == InventoryAction.COLLECT_TO_CURSOR) {
            event.isCancelled = true
            return
        }
        if (sellGUI.isConfirmMode()) {
            val action = currentItem?.itemMeta?.persistentDataContainer?.get(NamespacedKey(main, "guiAction"), PersistentDataType.STRING)
            val isConfirmButton = isGUIControlItem(currentItem) && action == "confirm"
            if (!isConfirmButton) {
                if (isGUIControlItem(currentItem) || isCustomMenuItem(currentItem)) {
                    event.isCancelled = true
                    return
                }
                sellGUI.setSellItem()
            }
        }
        when {
            isGUIControlItem(currentItem) -> {
                event.isCancelled = true
                when (currentItem?.itemMeta?.persistentDataContainer?.get(NamespacedKey(main, "guiAction"), PersistentDataType.STRING)) {
                    "confirm" -> {
                        if (sellGUI.hasUnevaluatedItems()) {
                            player.closeInventory()
                            val message = main.getMessagesConfig().getString("messages.evaluation-required", "&cYou have items that need to be evaluated before selling.")
                            player.sendMessage(ChatColor.translateAlternateColorCodes('&', message.orEmpty()))
                            return
                        }
                        SoundHandler.playConfigSound(player, "sounds.ui.confirm")
                        sellGUI.returnInvalidItems()
                        sellGUI.sellItems(sellGUI.getMenu())
                    }
                    "sell" -> {
                        sellGUI.returnInvalidItems()
                        if (sellGUI.getTotal(sellGUI.getMenu()) <= 0) {
                            SoundHandler.playConfigSound(player, "sounds.feedback.fail")
                            return
                        }
                        SoundHandler.playUIClick(player)
                        sellGUI.updateButtonState()
                    }
                }
            }
            isCustomMenuItem(currentItem) -> {
                event.isCancelled = true
                handleCustomMenuItemClick(player, currentItem!!)
            }
            else -> scheduleMenuRefresh(player, sellGUI)
        }
    }

    @EventHandler(ignoreCancelled = true)
    fun onPlayerDropItem(event: PlayerDropItemEvent) {
        if (main.config.getBoolean("general.add-worth-lore", false)) Bukkit.getScheduler().runTaskLater(main, event.player::updateInventory, 1L)
    }

    private fun scheduleMenuRefresh(player: Player, sellGUI: SellGUI) {
        Bukkit.getScheduler().runTaskLater(main, Runnable {
            SellCommand.getSellGUI(player)?.takeIf { it === sellGUI }?.let {
                it.returnInvalidItems()
                it.updateSellItemTotal()
            }
        }, 1L)
    }

    private fun isGUIControlItem(item: ItemStack?): Boolean = item?.takeIf { it.hasItemMeta() }?.itemMeta
        ?.persistentDataContainer?.has(NamespacedKey(main, "sellgui"), PersistentDataType.BYTE) == true

    private fun handleCustomMenuItemClick(player: Player, item: ItemStack) {
        val data = item.itemMeta.persistentDataContainer
        val commands = data.get(NamespacedKey(main, "custom-menu-item"), PersistentDataType.STRING)
        val sender = data.get(NamespacedKey(main, "custom-menu-item-sender"), PersistentDataType.STRING)
        val closeMenu = data.get(NamespacedKey(main, "custom-menu-item-close-menu"), PersistentDataType.BYTE)
        if (commands.isNullOrEmpty()) return
        if (closeMenu == 1.toByte()) {
            player.closeInventory()
            Bukkit.getScheduler().runTask(main, Runnable { executeCustomMenuCommands(player, commands, sender) })
        } else {
            executeCustomMenuCommands(player, commands, sender)
        }
    }

    private fun executeCustomMenuCommands(player: Player, commands: String, sender: String?) {
        commands.split(';').map(String::trim).filter(String::isNotEmpty).forEach { command ->
            val finalCommand = command.removePrefix("/")
            when (sender?.lowercase() ?: "console") {
                "player" -> player.performCommand(finalCommand)
                "op" -> {
                    val wasOp = player.isOp
                    try {
                        player.isOp = true
                        player.performCommand(finalCommand)
                    } finally {
                        player.isOp = wasOp
                    }
                }
                else -> Bukkit.dispatchCommand(Bukkit.getConsoleSender(), finalCommand)
            }
        }
    }

    private fun dropItems(inventory: Inventory, player: Player) {
        inventory.contents.forEach { item ->
            if (item == null || item.type == Material.AIR || isGUIControlItem(item) || isCustomMenuItem(item)) return@forEach
            if (main.config.getBoolean("drop-items-on-close", false)) {
                player.world.dropItem(player.location, item)
            } else {
                player.inventory.addItem(item).values.forEach { player.world.dropItem(player.location, it) }
            }
        }
    }

    private fun isCustomMenuItem(item: ItemStack?): Boolean = item?.takeIf { it.hasItemMeta() }?.itemMeta
        ?.persistentDataContainer?.has(NamespacedKey(main, "custom-menu-item"), PersistentDataType.STRING) == true
}
